superduperdb/components/graph.py

```diff
@@ -0,0 +1,196 @@
+from inspect import signature
+
+import networkx as nx
+
+from superduperdb.components.component import Component
+from superduperdb.components.model import Model
+
+
+class Key(str):
+    ...
+
+
+class KeyStore:
+    def __init__(self, model, params):
+        self.model = model
+        self._params = params
+        for param in params:
+            setattr(self, param, f'{model}.{param}')
+
+
+class GraphModel:
+    def __init__(self, model: 'Model'):
+        self.identifier = model.identifier
+        self.output = f'{model.identifier}.output'
+        self.forward_method = model.forward_method
+        self.object = model.object.artifact
+
+        params = self.set_input(model.object.artifact)
+        self.input = KeyStore(model.identifier, params)
+
+    def set_input(self, object):
+        predict = getattr(object, self.forward_method)
+
+        sig = signature(predict)
+        return list(sig.parameters.keys())
+
+
+class Node:
+    @staticmethod
+    def get_key(name):
+        key = '_base'
+        if name == '_Graph_input':
+            key = '_base'
+        else:
+            keys = name.split('.')
+            if len(keys) == 1:
+                key = '_base'
+            else:
+                key = keys[-1]
+        return key
+
+    def __init__(self, name, db=None):
+        self.name = name.split('.')[0]
+        key = name.split('.')[-1]
+        self.key = key if key else '_base'
+
+        self.model = GraphModel(db.load(name))
+        self._output = None
+
+    @property
+    def output(self):
+        return self._output
+
+    @output.setter
+    def output(self, output):
+        self._output = output
+
+
+class Graph(Component):
+    db: 'Datalayer'
+
+    def __post_init__(self):
+        self.input = '_Graph_input'
+        self.G = nx.DiGraph()
+        self._key_store = {}
+        self.nodes = {}
+        self._node_output_cache = {}
+
+    def add_edge(self, u, v):
+        v_key = None
+
+        if isinstance(v, Model):
+            v_key = '_base'
+            v = f'{v.identifier}.{v_key}'
+
+        elif isinstance(u, Model):
+            u = u.identifier
```

Comment:
How about `v` is Model and `u` is Model at the same time?

Reply:
Please check new interface, check the test cases